indico/modules/rb_new/event/fields.py

```diff
@@ -0,0 +1,79 @@
+# published by the Free Software Foundation; either version 3 of the
+# License, or (at your option) any later version.
+#
+# Indico is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+from wtforms.ext.sqlalchemy.fields import QuerySelectField
+
+from indico.util.i18n import _
+from indico.web.flask.util import url_for
+from indico.web.forms.widgets import SelectizeWidget
+
+
+class LinkedObjectField(QuerySelectField):
+
+    widget = SelectizeWidget(allow_by_id=True, search_field='title', label_field='full_title', preload=True,
+                             search_method='POST', inline_js=True)
+
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault('allow_blank', True)
+        kwargs['get_label'] = lambda a: '#{}: {}'.format(a.friendly_id, a.title)
+        self.ajax_endpoint = kwargs.pop('ajax_endpoint')
+        super(LinkedObjectField, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def _serialize_linked_object(cls, obj):
+        raise NotImplementedError
+
+    def _value(self):
+        return self._serialize_linked_object(self.data) if self.data else None
+
+    @property
+    def event(self):
+        # This cannot be accessed in __init__ since `get_form` is set
+        # afterwards (when the field gets bound to its form) so we
+        # need to access it through a property instead.
+        return self.get_form().event
+
+    @property
+    def search_url(self):
+        return url_for(self.ajax_endpoint, self.event)
+
+
+class ContributionField(LinkedObjectField):
+    """A selectize-based field to select a contribution that has no reservation yet."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault('render_kw', {}).setdefault('placeholder', _('Enter contribution title or #id'))
+        super(ContributionField, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def _serialize_linked_object(cls, contrib):
+        return {'id': contrib.id, 'friendly_id': contrib.friendly_id, 'title': contrib.title,
+                'full_title': '#{}: {}'.format(contrib.friendly_id, contrib.title)}
+
+
+class SessionBlockField(LinkedObjectField):
+    """A selectize-based field to select a session block that has no reservation yet."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault('render_kw', {}).setdefault('placeholder', _('Enter session block title or #id'))
+        super(SessionBlockField, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def _serialize_linked_object(cls, session_block):
+        return {'id': session_block.id, 'friendly_id': session_block.friendly_id, 'title': session_block.title,
```

Comment:
session blocks don't have a friendly ID and their title is optional -> use its `.full_title` instead.
since you can easily have multiple session blocks with the exact same `full_title` (multiple unnamed blocks in the same session), I think the date/time of it needs to be included as well